inc/basics.h

```diff
@@ -67,9 +67,9 @@ typedef uint32 unichar_t;
 /* A macro to print a string for debug purposes
  */
 #ifndef NDEBUG
-#define TRACE(...) printf(__VA_ARGS__)
+#define TRACE(...) fprintf(stderr, __VA_ARGS__)
 #else
-#define TRACE(...)
+#define TRACE(...) while(0)
```

Comment:
I don't understand how this even compiles (I assumed you'd want `do {} while (0)`), but it does, so this looks fine to me now :-)

Reply:
```
TRACE("u2utf8_copyn: truncated on invalid char 0x%x\n", ubuf[-1]);
```
(for example) is replaced by:
```
while(0);
```
which one could read as:
```
while (false)
    ;
```

Reply:
Oh, I see, okay, the semicolon makes it work. I thought a statement was needed, but I just tested it, and `if(true);` also is a no-op. Neat!

Reply:
Various different conventions turn up for this but they're often partly intended to "eat" the trailing semicolon on the TRACE or ASSERT or whatever, either because leaving it would be syntactically illegal or (more commonly) to avoid lint or compiler warnings.